Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md

```diff
@@ -617,9 +617,66 @@ However, this is not a concern on iOS. As discussed in the section [Testing on t
+```
+
+After the obfuscation we can observe that the symbol’s name is no longer meaningful as shown on the listing below.
+
+```assembly
+__T07DVIA_v232zNNtWKQptikYUBNBgfFVMjSkvRdhhnbyyFySbyypF:
+stp        x22, x21, [sp, #-0x30]!
+mov        rbp, rsp
+```
+
+Nevertheless, the binary’s code is not modified, so an attacker can still retrieve the logic of the algorithm by reading the disassembled version of the function and understand its purpose.
+
+#### Instructions substitution
+
+This technique replaces standard binary operators like addition or subtraction with more complex representations. For example an addition `x = a + b` can be represented as `x = -(-a) - (-b)`. However, using the same replacement representation could be easily reversed, so it is recommended to add multiple substitution techniques for a single case and introduce a random factor. Nevertheless, this technique is vulnerable to de-obfuscation, but it can still be time consuming depending on the complexity and depth of the substitutions.
+
+#### Control Flow Flattening
+
+Control Flow Flattening replaces original code with a more complex representation. The transformation breaks the body of the function into basic blocks and puts them all inside a single infinite loop with a switch statement that controls the program flow. This makes the program flow significantly harder to follow because the natural conditional constructs that made the code easier to read are now gone.
+
+![control-flow-flattening](./Images/Chapters/0x06j/control-flow-flattening.png)
+
+The image shows how the control-flow flattening alters code ([Obfuscating C++ programs via control flow flattening](http://ac.inf.elte.hu/Vol_030_2009/003.pdf))
+
+#### Dead code injection
+
+This technique makes the program's control-flow more complex by injecting dead code into the program. Dead code is a stub of code that doesn’t affect the original program’s behaviour but increases the overhead for the reverse engineering process.
+
+#### String encryption
+
+Applications are often compiled with hardcoded keys, licences, tokens and URL endpoints. By default, all of them are stored in plaintext in the data section of an application’s binary. String encryption aims to replace these values with their encrypted forms and inject stubs of code into the program that decrypts data before it is used by the program.
+
+#### Recommended tools
+
+- [SwiftShield](https://github.com/rockbruno/swiftshield) can be used to obfuscate binary's symbols. It reads the source code of the Xcode project and replaces all names of classes, methods and fields with random values before the compiler is used.
+- [obfuscator-llvm](https://github.com/obfuscator-llvm) operates on the Intermediate Representation instead of the the source code. It can be used for symbols obfuscation, string encryption and control-flow flattening. Since it's based on IR, it can hide out significantly more information about the application as compared to SwiftShield.
```

Comment:
```suggestion
- [obfuscator-llvm](https://github.com/obfuscator-llvm) operates on the Intermediate Representation (IR) instead of the source code. It can be used for symbol obfuscation, string encryption and control flow flattening. Since it's based on IR, it can hide out significantly more information about the application as compared to SwiftShield.
```